File: code/backend/app/routes/categories.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session
from app.db.database import get_db
from app.dependencies.auth import get_current_user
from app.models.user import User
from app.schemas.category import CategoryResponse, CategoryCreate, CategoryUpdate, AssignCategoryRequest
from app.schemas.recipe import RecipeResponse
from app.services import category_service


router = APIRouter(prefix="/categories", tags=["categories"])


@router.post("", response_model=CategoryResponse, status_code=status.HTTP_201_CREATED)
def create_category(
    data: CategoryCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return category_service.create_category(db, data.name)


@router.get("", response_model=list[CategoryResponse])
def get_categories(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return category_service.get_categories(db)


@router.patch("/{category_id}", response_model=CategoryResponse)
def update_category(
    category_id: int,
    data: CategoryUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return category_service.update_category(db, category_id, data.name)


@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_category(
    category_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    category_service.delete_category(db, category_id)


@router.post("/assign", response_model=list[RecipeResponse])
def assign_category(
    data: AssignCategoryRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return category_service.assign_recipes_to_category(db, data.recipe_ids, data.category_id, current_user.id)

File: code/frontend/src/api/categories.ts
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export interface Category {
  id: number;
  name: string;
  created_at: string;
}

export interface CategoryCreateRequest {
  name: string;
}

export interface CategoryUpdateRequest {
  name: string;
}

export interface AssignCategoryRequest {
  recipe_ids: number[];
  category_id: number | null;
}

const categoriesApi = {
  getCategories: async (token: string): Promise<Category[]> => {
    const response = await axios.get(`${API_URL}/categories`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    return response.data;
  },

  createCategory: async (data: CategoryCreateRequest, token: string): Promise<Category> => {
    const response = await axios.post(`${API_URL}/categories`, data, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    return response.data;
  },

  updateCategory: async (
    categoryId: number,
    data: CategoryUpdateRequest,
    token: string
  ): Promise<Category> => {
    const response = await axios.patch(`${API_URL}/categories/${categoryId}`, data, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    return response.data;
  },

  deleteCategory: async (categoryId: number, token: string): Promise<void> => {
    await axios.delete(`${API_URL}/categories/${categoryId}`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
  },

  assignCategories: async (data: AssignCategoryRequest, token: string): Promise<any> => {
    const response = await axios.post(`${API_URL}/categories/assign`, data, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    return response.data;
  },
};

export default categoriesApi;
